Blog post for how Pulumi can be used to automate member account creation in AWS Organizations (#2397)
pulumi/pulumi-hugo: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md

---
title: "Organizing Aws Accounts With Pulumi"
date: 2023-01-16
meta_desc: Learn how you can use Pulumi to manage your AWS Organizations hierarchy and automatically create member accounts.
meta_image: meta.png
authors:
    - praneet-loke
tags:
    - aws-organizations
---

In an Enterprise organization, an IT Self Service Vending Machine allows employees to quickly and easily request and receive access to pre-approved cloud resources. Behind the scenes, Pulumi programs may orchestrate any of the requisite resources. We will look at an example of using Pulumi to create an AWS child account, within an AWS Organization.

<!--more-->

> One might envision a developer requesting a new AWS account to be added entirely through a Pull Request in a git repository either for their own use or on behalf of a team. This scenario is not just for enterprises, however. Teams of any size should organize their AWS accounts using the least-privileged access principle.

Author: I have to admit I am not very fond of this opener a whole lot.

Reviewer: I think this is fine!

---
title: "Organizing Aws Accounts With Pulumi"

> date: 2023-01-16

Author: I picked next Monday as the date but based on when this can be reviewed/merged, I'll update this.

meta_desc: Learn how you can use Pulumi to manage your AWS Organizations hierarchy and automatically create member accounts.
meta_image: meta.png
authors:
    - praneet-loke
tags:
    - aws-organizations
---

> In an Enterprise organization, an IT Self Service Vending Machine allows employees to quickly and easily request and receive access to pre-approved cloud resources. Behind the scenes, Pulumi programs may orchestrate any of the requisite resources. We will look at an example of using Pulumi to create an AWS child account, within an AWS Organization.

Reviewer 2: ```suggestion
In an enterprise organization, an IT self-service "vending machine" allows employees to quickly and easily request and receive access to pre-approved cloud resources. Behind the scenes, Pulumi programs may orchestrate any of the requisite resources. We will look at an example of using Pulumi to create an AWS child account, within an AWS Organization.
```

<!--more-->

One might envision a developer requesting a new AWS account to be added entirely through a Pull Request in a git repository either for their own use or on behalf of a team. This scenario is not just for enterprises, however. Teams of any size should organize their AWS accounts using the least-privileged access principle.

Here’s the Organization structure we are looking to build with Pulumi. It is intentionally simple to demonstrate how teams can construct the structure that works best for them or follow AWS’ recommendations on this topic. Depending on your team size you may want to keep this simple or you may need something that can handle multiple teams each with multiple accounts. The decision is yours.

> ![Organization Structure](org-structure.png)

Author: Done. I've updated the paragraph before this slightly to do just that. Let me know what you think.

For the above organization structure, we want the following operational model:

> 1. Developer accounts will be provisioned (either manually or automatically through SSO depending on your HR systems) in the root account without any permissions other than being able to assume a specific role in the target accounts that have a trust boundary with the root account.

Author: I'll reword "developer accounts" here. What I meant was the actual users of the AWS account who are developers. Not the development AWS accounts. Haha! I see how confusing this sentence is. Even I did a double-take reading this. 😄

**Note**: If you are using one of the supported CI/CD services you might be able to use federated credentials using OIDC. So if you are going to use that auth mechanism, you can skip creation of the fixed automation user account since you will not need its
credentials.

## Organizational Unit (OU)

An org unit can represent anything. In our example it represents the development team. You could also have multiple OUs and even nest them to create sub-hierarchies, depending on your use-case.

The `organizations` module in the AWS provider package contains the resources we need in order to create an OU and an account.

```typescript
const devOrgUnit = new aws.organizations.OrganizationalUnit("orgUnit", {

>    parentId: organization.then((o) => o.roots[0].id),

Author: Yes definitely. The `organization` variable's value depends on whether or not the user has already created the organization in their root account. But I'll include that snippet.

>    parentId: organization.then((o) => o.roots[0].id),

Author: Ah yes. Looks like all code snippets in this post are somehow using three-space indentation. I must have miscounted them 😄 as I had to adjust these manually after a copy/paste from a Google Doc. I'll fix them.

       closeOnDeletion: true,
   },
   { protect: true }
);
```

{{% notes type="info" %}}
An important note about the `Account` resource and destroying them. Notice how we set the `closeOnDeletion` to `true`. Closing accounts on deletion will only succeed if you have not exceeded the quotas set by AWS for account closure. So set that flag accordingly. If you expect to be closing many accounts frequently (perhaps for an ephemeral use-case), you may want to set the flag to `false` so you can close them manually or have some other sort of asynchronous automation do it for you. Otherwise, your Pulumi execution will fail when AWS returns an error code due to the quota being exceeded.
{{% /notes %}}

## Developers’ Access

We’ll create an IAM group for developers and assign a group policy that allows its members to perform the `sts:AssumeRole` on any AWS account but only using a specific role.

```typescript
const devGroup = new aws.iam.Group("developers", {
   name: "developers",
});

new aws.iam.GroupPolicy("developersGroupPolicy", {
   group: devGroup.name,
   policy: {
       Statement: [
           {
               Effect: "Allow",
               Action: "sts:AssumeRole",

>                Resource: `arn:aws:iam::*:role/DeveloperRole`,

Reviewer: Also here -- might be good to just mention briefly that these (this one and `AutomationRole`) are AWS built-ins.

Author: This role and the `AutomationRole` are roles that the app creates. In the actual example I have these in two variables. I replaced the variable references with hard-coded strings for the sake of keeping things simple in the code snippet in this post. The relevant lines in the actual example: https://github.com/pulumi/examples/pull/1330/files#diff-b8897f56ce67084e1d7fd95fb1af74032ef404fee05abbb4ab013340f384b6a6R18-R19

Author: But I'll add a comment that these are roles created by the component resource for account permissions.

           },
       ],
       Version: "2012-10-17",
   },
});
```

> And create the IAM user that we’ll use for automation purposes. Again, you don’t need to do this if you plan to use OIDC federation for your CI/CD service.

Reviewer 2: ```suggestion
Next, we'll create the IAM user that we’ll use for automation purposes. Again, you don’t need to do this if you plan to use OIDC federation for your CI/CD service.
```

```typescript
const devAccountProvider = new aws.Provider("devAccountProvider", {
   allowedAccountIds: [devAccount.id],
   assumeRole: {
       roleArn: pulumi.interpolate`arn:aws:iam::${devAccount.id}:role/OrganizationalAccountAccessRole`,
   },
});
```

Now, whenever we want to modify resources in the target account in this app, we’ll use that provider. We’ll pass this provider to the component mentioned above, so that the roles and policies created by that component are performed using this provider instance. The component resource code is not included in this post for brevity. It provides a nice way to encapsulate the roles and policies we want to bootstrap our brand new account with. Of course, you can customize those roles and what permissions they get based on your use-case.

```typescript
const devAccountPermissions = new AccountPermissions(
   "devAccountPermissions",
   {
       automationUser,
       managementAccountId: pulumi.Output.create(
           aws.getCallerIdentity().then((i) => i.accountId)
       ),
   },
   { provider: devAccountProvider }
);
```

> ## Backup Policies

Reviewer: This is cool, but feels somewhat orthogonal to the content of the post, which is about account creation. Totally fine to leave it in -- I just figured I'd mention it as something to consider if you were thinking about making any revisions.

(Out of curiosity, though, if you do leave it in, it might be good to mention how it works, and sprinkle in a link or two to the AWS docs. For example, I sort of assumed I'd see something about what kinds of "things" would be backed up (DB, etc., as you mention), but that doesn't seem to be called out in the code.)

Reviewer 2: If I'm understanding correctly, I _think_ the idea here is to also show the "other things" that need to be done when creating accounts within an AWS Organization. If that's the case, I'd suggest some sort of transition here ("In addition to creating and deleting accounts, most teams will also want to make sure that resources are backed up, tagging is enforced...").

Author: @cnunciato

> This is cool, but feels somewhat orthogonal to the content of the post, which is about account creation. Totally fine to leave it in -- I just figured I'd mention it as something to consider if you were thinking about making any revisions.

It's basically what @scottslowe mentions in his comment above. I think the suggestion to add a transition is a good one. I'll update this.

Author: I've updated this section and the next. Let me know what you think.

## Tagging Policies

In keeping with our general theme here, we’ll encapsulate our tagging policy with yet another component resource so that we can easily include future org units. Similar to the backup policy syntax, the tag policy syntax uses JSON with some special placeholder variables.

Let’s take a look at the consumer point-of-view for this component. We want users of this component to pass in a list of allowed cost centers for a certain org unit. Say, if a team wants to have multiple org units and therefore different cost centers for each of them, this model allows those teams to use the `TagPolicies` component for each org unit separately and assign the allowed cost center values accordingly.

```typescript
const tagPolicies = new TagPolicies("tagPolicies", {
   costCenters: [
       {
           allowedCostCenters: ["Development", "Testing"],
           ou: devOrgUnit,
       },
   ],
   orgId: organization.then((o) => o.roots[0].id),
});
```

The tag policy for our example looks like this:

```typescript
{
   type: "TAG_POLICY",
   content: JSON.stringify({
       tags: {
           CostCenter: {
               tag_value: {

>                    "@@assign": costCenter.allowedCostCenters,

Reviewer: Might be good to explain what `costCenter` refers to.

               },
           },

           // Add tags that are enforceable only to the OU.
           Owner: {
               tag_key: {
                   "@@assign": "Owner",
                   // Don't allow any child policies from changing the key of this tag.
                   "@@operators_allowed_for_child_policies": [
                       "@@none",
                   ],
               },
           },
       },
   }),
}
```

And that’s about it as far as setting up the simple org unit and member account structure shown at the beginning of this post.
View the [full source-code](https://github.com/pulumi/examples/tree/master/aws-ts-organizations) for the component resources in the `pulumi/examples` repository.

> This post demonstrated how you could use the Pulumi programming model to encapsulate and account for the hierarchical organization structure that your team desires or is required to have for any number of reasons. While an enterprise organization might be interested in this sort of a setup to ensure multiple teams in the organization operate in a similar way, smaller teams should also strongly consider organizing their AWS accounts in a way that enforces the principle of least privileged access. Especially, if you are looking to get SOC2 certified and need to ensure strict isolation of resources that hold any kind of real-world customer data.

Reviewer: ```suggestion
This post demonstrates how you could use the Pulumi programming model to encapsulate and account for the hierarchical organization structure that your team desires or is required to have for any number of reasons. While an enterprise organization might be interested in this sort of a setup to ensure multiple teams in the organization operate in a similar way, smaller teams should also strongly consider organizing their AWS accounts in a way that enforces the principle of least privileged access. Especially, if you are looking to get SOC2 certified and need to ensure strict isolation of resources that hold any kind of real-world customer data.
```

> ## Next Steps

Reviewer: I'd really like to see what this looks like when it runs! Partly because as you suggest, many users won't have the appropriate permissions to actually deploy it. If there were some way to show what happens after a successful run, it'd really bring it all home I think. Up to you though.

Author: Yes definitely agree. Do you think a GIF or a screenshot of the Pulumi plan would be good here?

Reviewer 2: An animated GIF that shows part of the `pulumi up` output would be cool, I think.

Author: I've expanded the outro a bit more along with a gif. Let me know what you think. If you would like to see just those changes, it would be commit 44254a42d6429d12975fe39ac3478610a80df619.

---
title: "Organizing AWS Accounts With Pulumi"
date: 2023-02-06
meta_desc: Learn how you can use Pulumi to manage your AWS Organizations hierarchy and automatically create member accounts.
meta_image: meta.png
authors:
    - praneet-loke
tags:

>     - aws-organizations

Reviewer: ```suggestion
    - aws
    - enterprise
    - components
```
